File: Net/Aircompany/Planes/MilitaryPlane.cs
using Aircompany.Models;

namespace Aircompany.Planes
{
    public class MilitaryPlane : Plane
    {
        public MilitaryType militaryType;

        public MilitaryPlane(string model, int maxSpeed, int maxFlightDistance, int maxLoadCapacity, MilitaryType type)
            : base(model, maxSpeed, maxFlightDistance, maxLoadCapacity)
        {
            this.militaryType = type;
        }

        public override bool Equals(object obj)
        {
            var plane = obj as MilitaryPlane;
            return plane != null &&
                   base.Equals(obj) &&
                   militaryType == plane.militaryType;
        }

        public override int GetHashCode()
        {
            var hashCode = 1701194404;
            hashCode = hashCode * -1521134295 + base.GetHashCode();
            hashCode = hashCode * -1521134295 + militaryType.GetHashCode();
            return hashCode;
        }

        public MilitaryType GetMilitaryPlaneType()
        {
            return militaryType;
        }


        public override string ToString()
        {
            return base.ToString().Replace("}",
                    ", type=" + militaryType +
                    '}');
        }        
    }
}

File: Java/src/main/java/Planes/MilitaryPlane.java
package Planes;

import models.MilitaryType;

import java.util.Objects;

public class MilitaryPlane extends Plane {
    private final MilitaryType type;

    private MilitaryPlane(Builder builder) {
        super(builder.model, builder.maxSpeed, builder.maxFlightDistance, 0);
        this.type = builder.type;
    }

    public static class Builder {
        private final String model;
        private final int maxSpeed;
        private final int maxFlightDistance;
        private MilitaryType type;

        public Builder(String model, int maxSpeed, int maxFlightDistance) {
            this.model = model;
            this.maxSpeed = maxSpeed;
            this.maxFlightDistance = maxFlightDistance;
        }

        public Builder type(MilitaryType type) {
            this.type = type;
            return this;
        }

        public MilitaryPlane build() {
            return new MilitaryPlane(this);
        }
    }

    public static Builder builder(String model, int maxSpeed, int maxFlightDistance) {
        return new Builder(model, maxSpeed, maxFlightDistance);
    }

    public MilitaryType getType() {
        return type;
    }

    @Override
    public String toString() {
        return getClass().getSimpleName() + "{" +
                "model='" + super.getModel() + '\'' +
                ", maxSpeed=" + super.getMaxSpeed() +
                ", maxFlightDistance=" + super.getMaxFlightDistance() +
                ", maxLoadCapacity=" + super.getMaxLoadCapacity() +
                ", type=" + type +
                '}';
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof MilitaryPlane)) return false;
        if (!super.equals(o)) return false;
        MilitaryPlane that = (MilitaryPlane) o;
        return type == that.type;
    }

    @Override
    public int hashCode() {
        return Objects.hash(super.hashCode(), type);
    }
}
